File: frame/src/main/java/com/dreampany/frame/data/misc/StoreMapper.java
package com.dreampany.frame.data.misc;

import com.dreampany.frame.data.model.Store;
import com.dreampany.frame.misc.SmartCache;
import com.dreampany.frame.misc.SmartMap;
import com.dreampany.frame.misc.StoreAnnote;

import javax.inject.Inject;

/**
 * Created by Hawladar Roman on 8/9/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
public class StoreMapper {

    //keeps only LRU
    private final SmartMap<Long, Store> map;
    private final SmartCache<Long, Store> cache;

    @Inject
    StoreMapper(@StoreAnnote SmartMap<Long, Store> map,
                @StoreAnnote SmartCache<Long, Store> cache) {
        this.map = map;
        this.cache = cache;
    }

    public boolean isExists(long id) {
        return map.contains(id);
    }

    public boolean isExists(long id, String type, String subtype) {
        if (isExists(id)) {
            Store item = getItem(id);
            return item.hasProperty(type, subtype);
        }
        return false;
    }

    public void putItem(Store item) {
        map.put(item.getId(), item);
    }

    public Store getItem(long id) {
        return map.get(id);
    }

    public Store getItem(long id, String type, String subtype) {
        if (isExists(id, type, subtype)) {
            return getItem(id);
        }
        return null;
    }
}

File: common/src/main/java/com/dreampany/common/data/source/mapper/StoreMapper.kt
package com.dreampany.common.data.source.mapper

import com.dreampany.common.data.model.Store
import com.google.common.collect.Maps
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by roman on 1/5/20
 * Copyright (c) 2020 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Singleton
class StoreMapper
@Inject constructor() {

    private val stores: MutableMap<String, Store>

    init {
        stores = Maps.newConcurrentMap()
    }

    fun isExists(id: String): Boolean {
        return stores.contains(id)
    }

    fun isExists(id: String, type: String, subtype: String, state: String): Boolean {
        if (isExists(id)) {
            val item = getItem(id)
            return item?.hasProperty(type, subtype, state) ?: false
        }
        return false
    }

    fun putItem(item: Store) {
        stores.put(item.id, item)
    }

    fun getItem(id: String): Store? {
        return stores.get(id)
    }

    fun getItem(id: String, type: String, subtype: String, state: String, extra: String? = null): Store? {
        return if (isExists(id, type, subtype, state)) getItem(id)
        else Store(
            id = id,
            type = type,
            subtype = subtype,
            state = state,
            extra = extra
        )
    }
}